artifact/v3/header/token.hpp

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_ARTIFACT_V3_HEADER_TOKEN_HPP
+#define MENDER_ARTIFACT_V3_HEADER_TOKEN_HPP
+
+
+#include <unordered_map>
+#include <string>
+#include <regex>
+
+#include <common/log.hpp>
+#include <common/expected.hpp>
+
+const std::regex artifact_script_regexp {
+	"scripts/Artifact(Install|Reboot|Rollback|RollbackReboot|Failure)_(Enter|Leave)_[0-9][0-9](_[^[:space:]]+)?",
+	std::regex_constants::ECMAScript};
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace header {
+namespace token {
+
+
+using namespace std;
+
+namespace tar = mender::tar;
+
+namespace log = mender::common::log;
+namespace error = mender::common::error;
+namespace expected = mender::common::expected;
+
+enum class Type {
+	Uninitialized = 0,
+	EOFToken,
+	Unrecognized,
+	HeaderInfo,
+	ArtifactScripts,
+	ArtifactHeaderTypeInfo,
+	ArtifactHeaderMetaData,
+};
+
+const unordered_map<const Type, const string> type_map {
+	{Type::Uninitialized, "Uninitialized"},
+	{Type::EOFToken, "EOF"},
+	{Type::Unrecognized, "Unrecognized"},
+	{Type::HeaderInfo, "header-info"},
+	{Type::ArtifactScripts, "artifact-state-scripts"},
+	{Type::ArtifactHeaderTypeInfo, "header-type-info"},
+	{Type::ArtifactHeaderMetaData, "header-meta-data"},
+};
+
+class Token {
+public:
+	Type type;
+	// Poor mans optional
+	shared_ptr<tar::Entry> value;
+
+public:
+	Token() :
+		type {Type::Uninitialized} {
+	}
+	Token(Type t) :
+		type {t} {
+	}
+	Token(const string &type_name, tar::Entry &entry) :
+		type {StringToType(type_name)},
+		value {make_shared<tar::Entry>(entry)} {
+	}
+
+	const string TypeToString() const {
+		return type_map.at(type);
+	}
+
+private:
+	Type StringToType(const string &type_name) {
+		if (type_name.find("header-info") == 0) {
+			return Type::HeaderInfo;
+		}
+		if (std::regex_match(type_name, artifact_script_regexp)) {
+			return Type::ArtifactScripts;
+		}
+		if (std::regex_match(type_name, regex("headers/[0-9]{4}/type-info"))) {
+			return Type::ArtifactHeaderTypeInfo;
+		}
+		if (regex_match(type_name, regex("headers/[0-9]{4}/meta-data"))) {
+			return Type::ArtifactHeaderMetaData;
+		}
+		log::Error("Unrecognized token: " + type_name);
+		return Type::Unrecognized;
```

Comment:
This might not be the right place to handle it (and maybe you already have), but I'll just point it out: I remember that I fixed a bug in mender-artifact a long time ago where it couldn't read an artifact which was artificially put together using `tar`. While this is of course not common, it has its debugging uses and occasionally even in production. Notice the difference between them:

File list from mender-artifact:

```
headers/0000/type-info
```

File list from `tar`:

```
headers
headers/0000
headers/0000/type-info
```

In other words it contains the directories leading up to the file as well. This should be filtered out by looking at file type.

Reply:
I won't handle this here, and maybe we should discuss this a little more.

So in short this means that we will accept Artifacts which do not align with the format we have described, I know that the regular cli tools create this form of structure, but then we should update our documentation also, no?

Reply:
I wouldn't say that it doesn't align. It's just that the `tar` tool creates a slightly different internal structure compared to the Go tar writer, but it's still semantically valid. It doesn't really change anything, it's just the same thing represented a bit differently.

It should be pretty easy to filter out though, shouldn't it? Something like (pseudo code):

```
if (entry.type == directory) {
    if (entry.path is a prefix of any valid token) {
        ignore();
    } else {
        error("Forbidden entry");
    }
}
```

Reply:
Hmm, alright. But keeping it out of this PR

Reply:
Fair enough.